src/lib/core/mp_datetime.c

```diff
@@ -0,0 +1,194 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#include <limits.h>
+
+#include "msgpuck.h"
+#include "mp_datetime.h"
+#include "mp_extension_types.h"
+#include "mp_utils.h"
+
+/*
+  Datetime MessagePack serialization schema is MP_EXT (0xC7 for 1 byte length)
+  extension, which creates container of 1 to 3 integers.
+
+  +----+-----------+---+====~~~~~~~====+....~~~~~~~~~....+----~~~~~~~~----+....~~~~....+
+  |0xC7|len (uint8)| 4 | seconds (int) | tzoffset (int)  | tzindex (uint) | nsec (int) |
+  +----+-----------+---+====~~~~~~~====+....~~~~~~~~~....+----~~~~~~~~----+....~~~~....+
+
+  MessagePack extension MP_EXT (0xC7), after 1-byte length, contains:
+
+  - signed integer seconds part (required). Depending on the value of
+    seconds it may be from 1 to 8 bytes positive or negative integer number;
+
+  - [optional] fraction time in nanoseconds as unsigned integer.
+    If this value is 0 then it's not saved (unless there is offset field,
+    as below);
+
+  - [optional] timezone offset in minutes as signed integer.
+    If this field is 0 then it's not saved.
+ */
+
+
+#define check_secs(secs)                                \
+	assert((int64_t)(secs) <= MAX_EPOCH_SECS_VALUE);\
+	assert((int64_t)(secs) >= MIN_EPOCH_SECS_VALUE);
+
+#define check_nanosecs(nsec)      assert((nsec) < 1000000000);
+
+#define check_tz_offset(offset)       \
+	assert((offset) <= (12 * 60));\
+	assert((offset) >= (-12 * 60));
+
+static inline uint32_t
+mp_sizeof_datetime_raw(const struct datetime *date)
+{
+	uint32_t sz = 0;
+	if (date->epoch != 0 || date->tzoffset != 0 ||
+	    date->tzindex != 0 || date->nsec != 0) {
+		check_secs(date->epoch);
+		sz += mp_sizeof_xint(date->epoch);
+	}
+	if (date->tzoffset != 0 || date->tzindex != 0 || date->nsec != 0) {
+		check_tz_offset(date->tzoffset);
+		sz += mp_sizeof_xint(date->tzoffset);
+	}
+	if (date->tzindex != 0 || date->nsec != 0) {
+		sz += mp_sizeof_xint(date->tzindex);
+	}
+	if (date->nsec != 0) {
+		check_nanosecs(date->nsec);
+		sz += mp_sizeof_xint(date->nsec);
+	}
+	return sz;
+}
+
+uint32_t
+mp_sizeof_datetime(const struct datetime *date)
+{
+	return mp_sizeof_ext(mp_sizeof_datetime_raw(date));
+}
+
+struct datetime *
+tnt_datetime_unpack(const char **data, uint32_t len, struct datetime *date)
+{
+	const char *svp = *data;
+	memset(date, 0, sizeof(*date));
+
+	if (len <= 0)
+		return date;
```

Comment:
As i see for `decimal` and `uuid` in case of error we restore data pointer and return NULL
*data = svp;
return NULL

Reply:
Fixed as...
```diff
--- a/src/lib/core/mp_datetime.c
+++ b/src/lib/core/mp_datetime.c
@@ -56,10 +56,11 @@ mp_sizeof_datetime(const struct datetime *date)
 struct datetime *
 tnt_datetime_unpack(const char **data, uint32_t len, struct datetime *date)
 {
-	memset(date, 0, sizeof(*date));
-
 	if (len <= 0)
-		return date;
+		return NULL;
+
+	const char *const svp = *data;
+	memset(date, 0, sizeof(*date));
 
 	int64_t i_epoch;
 	memcpy(&i_epoch, *data, sizeof(i_epoch));
@@ -67,10 +68,13 @@ tnt_datetime_unpack(const char **data, uint32_t len, struct datetime *date)
 	*data += sizeof(i_epoch);
 	len -= sizeof(i_epoch);
 
-	if (len <= 0)
+	if (len == 0)
 		return date;
 
-	assert(len == SZ_TAIL);
+	if (len != SZ_TAIL) {
+		*data = svp;
+		return NULL;
+	}
 	memcpy(&date->nsec, *data, SZ_TAIL);
 	*data += SZ_TAIL;
```